Fix incorrect batch sizes in KFoldIterator

printomi opened the pull request:
## What changes were proposed in this pull request?
This PR will fix #6786 
[KFoldIterator](https://github.com/deeplearning4j/deeplearning4j/blob/release/1.0.0-beta3/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/iterator/KFoldIterator.java) has been modified following the [scikit-learn implementation](https://github.com/scikit-learn/scikit-learn/blob/55bf5d93e5674f13a1134d93a11fd0cd11aabcd1/sklearn/model_selection/_split.py#L438).

`N` number of samples are split into `k` batches. The first `(N%k)` batches contain `(N/k)+1` samples, while the remaining batches contain `(N/k)` samples. 
In case the number of samples (`N`) in the dataset is a multiple of `k`, all batches will contain `(N/k)` samples.

Visibility of private fields and methods in `KFoldIterator` has been changed to protected, so that this class can be extended in the future. (Stratified k-fold and leave-k-out cross-validation iterators would be nice.)

An unneccessary copy of the `DataSet` has been eliminated from `KFoldIterator` constructor. However, the `next()` method will maintain a copy in the fields `train` and `test`.

[Unit tests](https://github.com/deeplearning4j/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/dataset/KFoldIteratorTest.java) has been modified too: enhancements, improved documentation comments, and an added new test case.

## How was this patch tested?
Manual test run in the directory `deeplearning4j/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native` with the following command:
```
mvn -e test -Dtest=KFoldIteratorTest -P testresources
```

RobAltena commented:
This needs to be in the source code comments too. 
> I modified KFoldIterator following the scikit-learn implementation.

The 2nd file we should modify in this PR is the KFoldIteratorTest.java.

printomi commented:
Thanks for the comment, I will include a comment about this major change of implementation in the author field of the `KFoldIterator` JavaDoc.
I will also use your [minor maintenance changes](https://github.com/deeplearning4j/deeplearning4j/pull/6792) in the unit test update.

printomi commented:
Finally, I updated `KFoldIteratorTest`, and simplified KFoldIterator so that the main logic be more straightforward.
I see that I made a conflicting change: it is nothing breaking, but a function that is placed above an inner class. I will try to resolve it by placing back this function to it's original place in the file.

printomi commented:
I have just removed an unneccessary copy of the DataSet from KFoldIterator.
I think that I can't improve any more this PR, so a review would be great now. 
@RobAltena , could you check if the tests pass?

printomi commented:
Manual test is in progress. I have just merged upstream branch. Rebuilding libnd4j, then recompiling and running tests.

printomi commented:
All tests passed with latest code. I used the following command in the directory `deeplearning4j/nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native` to run the tests:
```
mvn -e test -Dtest=KFoldIteratorTest -P testresources
```

RobAltena commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/iterator/KFoldIterator.java:
```diff
@@ -24,43 +24,64 @@
+     * Create a k-fold cross-validation iterator given the dataset and k=10 train-test splits.
+     * N number of samples are split into k batches. The first (N%k) batches contain (N/k)+1 samples, while the remaining batches contain (N/k) samples. 
+     * In case the number of samples (N) in the dataset is a multiple of k, all batches will contain (N/k) samples.
+     * @param k number of folds (optional, defaults to 10)
```
There is no parameter k in this constructor.
